pygeos/strtree.py

```diff
@@ -303,3 +303,67 @@ def nearest_all(self, geometry, max_distance=None, return_distance=False):
             return self._tree.nearest_all(geometry, max_distance)
 
         return self._tree.nearest_all(geometry, max_distance)[0]
+
+    @requires_geos("3.10.0")
+    def dwithin(self, geometry, distance):
+        """Returns all combinations of each input geometry and geometries in the tree
+        where each tree geometry is within distance of each input geometry.
+
+        This returns an array with shape (2,n) where the subarrays correspond
+        to the indexes of the input geometries and indexes of the tree geometries
+        associated with each.  To generate an array of pairs of input geometry
+        index and tree geometry index, simply transpose the results.
+
+        Any geometry that is None or empty in the input geometries is omitted from
+        the output.
+
+        Parameters
+        ----------
+        geometry : Geometry or array_like
+            Input geometries to query the tree.  The envelope of each geometry
+            is automatically calculated for querying the tree.
+        distance : number or array_like
+            Distances within which to query the tree for each input geometry.
+            If array_like, shape must be broadcastable to shape of geometry.
+
+        Returns
+        -------
+        ndarray of shape (n,) if geometry is scalar, otherwise ndarray with shape (2, n)
+            The first subarray contains input geometry indexes.
+            The second subarray contains tree geometry indexes.
+
+        Examples
+        --------
+        >>> import pygeos
+        >>> tree = pygeos.STRtree(pygeos.points([[0, 0], [1, 1], [0, 4]]))
+        >>> tree.dwithin(pygeos.points([[0,1]]), 1.25).tolist()  # doctest: +SKIP
+        [[0, 0], [0, 1]]
+        >>> tree.dwithin(pygeos.points([[0,0]]), 0).tolist()  # doctest: +SKIP
+        [[0], [0]]
+        >>> tree.dwithin(pygeos.points([[0,0]]), np.nan).tolist()  # doctest: +SKIP
+        [[], []]
+        >>> tree.dwithin(pygeos.Geometry('POINT EMPTY'), 10).tolist()  # doctest: +SKIP
+        [[], []]
+        """
+
+        is_scalar = np.isscalar(geometry)
```

Comment:
This isn't working the way I expected.  I'm trying to test for singular geometries here (which may be `None`), but `np.isscalar` is false for these.

The goal was to try and make this behave like `query` for singular inputs, and `query_bulk` for array-like inputs.  Maybe we should always treat inputs as array-like?

Reply:
I think I am against this API anyway. The type of the return value should not depend on the inputs given to a function.

We already have a pattern estabilished for this (`query` /  `query_bulk`), I think the logical option would be  `dwithin` / `dwithin_bulk`

Reply:
One thought: I understand that the C implementation is separate, but we could integrate dwithin into query/query_bulk on the python side only?